src/bentoml/_internal/cloud/client.py

```diff
@@ -735,18 +734,20 @@ def tail_logs(
         namespace: str,
         pod_name: str,
         container_name: str = "main",
-        stop_event: threading.Event | None = None,
-    ) -> t.Generator[t.Tuple[str, WebSocketSession], None]:
+    ) -> tuple[t.Generator[str, None, None], t.Callable[[], None]]:
         url_ = urlparse(self.endpoint)
         scheme = "wss"
         if url_.scheme == "http":
             scheme = "ws"
         endpoint = f"{scheme}://{url_.netloc}"
 
-        with connect_ws(
-            url=f"{endpoint}/ws/v1/clusters/{cluster_name}/tail?{urlencode(dict(namespace=namespace, pod_name=pod_name))}",
-            client=self.session,
-        ) as ws:
+        with contextlib.ExitStack() as stack:
+            ws = stack.enter_context(
```

Comment:
Thank you, it turns out that there's no need to upgrade to 3.8+ to avoid nested contexts, happy!